docs/how-to/test-dataset-definition.md

```diff
@@ -0,0 +1,186 @@
+# How to test your dataset definition
+
+This guide shows you how to evaluate and validate the behaviour of your ehrQL query.
+The [assure](../../reference/cli/#assure) command works like a suite of unit tests for your ehrQL queries.
+
+You can write assurance tests that help you and others to understand and review the expected behaviour of your dataset definition.
+Assurance tests also provide confidence that existing functionality remains unchanged when reworking your dataset definition.
+
+In this guide we first set up an example dataset definition and then demonstrate the steps needed to test the behaviour of the ehrQL queries in the dataset definition.
+
+## Example dataset definition
+
+First, we're creating a dataset definition with the following specifications:
+
+- The population includes everyone above the age of 18 on the `index_date` (31st March 2023).
+- The dataset has the following three columns:
+    - `age`: Patient's age on the `index_date`.
+    - `has_asthma_med`: Boolean value specifying whether a patient has received an asthma medication before the `index_date`.
+    - `latest_asthma_med_date`: Date of the most recent prescription of an asthma medication before the `index_date`.
+
+```ehrql
+from ehrql import create_dataset
+from ehrql.tables.core import patients, medications
+
+asthma_codes = ["39113311000001107", "39113611000001102"]
+
+dataset = create_dataset()
+
+index_date = "2023-03-31"
+
+dataset.age = patients.age_on(index_date)
+dataset.define_population(dataset.age > 18)
+
+latest_asthma_med = (
+    medications.where(medications.dmd_code.is_in(asthma_codes))
+    .where(medications.date <= index_date)
+    .sort_by(medications.date)
+    .last_for_patient()
+)
+
+dataset.has_asthma_med = latest_asthma_med.exists_for_patient()
+dataset.latest_asthma_med_date = latest_asthma_med.date
+```
+
+## Specifying data and expectations
+
+Next, you need to provide (1) data for test patients and (2) define the expected outcomes for these patients after applying your ehrQL queries.
+Test data and expectations are defined in a nested dictionary called `patient_data`.
+The key in the outermost dictionary specifies the patient id.
+
+### Data for test patients
+
+In the example below we have created one test patient with the patient id `1` and specified more data for this patient in two lists called `patient` and `medications`.
+The names of the inner lists match the names of the tables you are using in your dataset definition.
+To explore how these tables are structured you can look at the column names in the [core schema](../../reference/schemas/core) documentation.
+You can also generate CSV files of these tables and add them to your project using the [dump-example-data](../../reference/cli/#dump-example-data) or [create-dummy-tables](../../reference/cli/#create-dummy-tables) commands.
```

Comment:
Does talking about dummy data complicate things here?

Reply:
probably yes, I think refering to the schema is helpful enough!